services/brig/src/Brig/API/OAuth.hs

```diff
@@ -0,0 +1,397 @@
+      cid = oriClientId info
+  void $ getOAuthClient uid cid >>= maybe (throwStd $ errorToWai @'OAuthClientNotFound) pure
+
+  unlessM (verifyClientSecret (ortrClientSecret req) cid) $ throwStd $ errorToWai @'OAuthInvalidClientCredentials
+  lift $ wrapClient $ deleteOAuthRefreshToken info
+
+--------------------------------------------------------------------------------
+
+getOAuthApplications :: UserId -> (Handler r) [OAuthApplication]
+getOAuthApplications uid = do
+  activeRefreshTokens <- lift $ wrapClient $ lookupOAuthRefreshTokens uid
+  nub . catMaybes <$> for activeRefreshTokens oauthApp
+  where
+    oauthApp :: OAuthRefreshTokenInfo -> (Handler r) (Maybe OAuthApplication)
+    oauthApp info = (OAuthApplication (oriClientId info) . ocName) <$$> getOAuthClient (oriUserId info) (oriClientId info)
+
+--------------------------------------------------------------------------------
+
+revokeOAuthAccountAccess :: UserId -> OAuthClientId -> (Handler r) ()
+revokeOAuthAccountAccess uid cid = do
+  rts <- lift $ wrapClient $ lookupOAuthRefreshTokens uid
+  for_ rts $ \rt -> when (oriClientId rt == cid) $ lift $ wrapClient $ deleteOAuthRefreshToken rt
+
+--------------------------------------------------------------------------------
+-- DB
+
+insertOAuthClient :: (MonadClient m) => OAuthClientId -> OAuthApplicationName -> RedirectUrl -> Password -> m ()
+insertOAuthClient cid name uri pw = retry x5 . write q $ params LocalQuorum (cid, name, uri, pw)
+  where
+    q :: PrepQuery W (OAuthClientId, OAuthApplicationName, RedirectUrl, Password) ()
+    q = "INSERT INTO oauth_client (id, name, redirect_uri, secret) VALUES (?, ?, ?, ?)"
+
+lookupOauthClient :: (MonadClient m) => OAuthClientId -> m (Maybe OAuthClient)
+lookupOauthClient cid = do
+  mNameUrl <- retry x5 . query1 q $ params LocalQuorum (Identity cid)
+  pure $ mNameUrl <&> uncurry (OAuthClient cid)
+  where
+    q :: PrepQuery R (Identity OAuthClientId) (OAuthApplicationName, RedirectUrl)
+    q = "SELECT name, redirect_uri FROM oauth_client WHERE id = ?"
+
+lookupOAuthClientSecret :: (MonadClient m) => OAuthClientId -> m (Maybe Password)
+lookupOAuthClientSecret cid = do
+  runIdentity <$$> retry x5 (query1 q (params LocalQuorum (Identity cid)))
+  where
+    q :: PrepQuery R (Identity OAuthClientId) (Identity Password)
+    q = "SELECT secret FROM oauth_client WHERE id = ?"
+
+insertOAuthAuthorizationCode :: (MonadClient m) => Word64 -> OAuthAuthorizationCode -> OAuthClientId -> UserId -> OAuthScopes -> RedirectUrl -> m ()
+insertOAuthAuthorizationCode ttl code cid uid scope uri = do
+  let cqlScope = C.Set (Set.toList (unOAuthScopes scope))
+  retry x5 . write q $ params LocalQuorum (code, cid, uid, cqlScope, uri)
+  where
+    q :: PrepQuery W (OAuthAuthorizationCode, OAuthClientId, UserId, C.Set OAuthScope, RedirectUrl) ()
+    q = fromString $ "INSERT INTO oauth_auth_code (code, client, user, scope, redirect_uri) VALUES (?, ?, ?, ?, ?) USING TTL " <> show ttl
```

Comment:
as below: should this be `scrypt`-ed?

Reply:
It can't be. We need to be able to look this up by the code.